src/PotatoTcp/Serialization/LengthPrefixedWireProtocol.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace PotatoTcp.Serialization
+{
+    public class LengthPrefixedWireProtocol : IWireProtocol
+    {
+        private BinaryFormatter _serializer = new BinaryFormatter();
+
+        private ILogger<LengthPrefixedWireProtocol> _logger;
+
+        public LengthPrefixedWireProtocol() : this(new NullLoggerFactory().CreateLogger<LengthPrefixedWireProtocol>()) { }
+
+        public LengthPrefixedWireProtocol(ILogger<LengthPrefixedWireProtocol> logger)
+        {
+            _logger = logger;
+        }
+
+        public object Deserialize(Stream stream)
+        {
+            try
+            {
+                var messageLength = ReadMessageLength(stream);
+
+                if (messageLength == 0)
+                    throw new InvalidOperationException("Received message of length 0.");
+
+                using (var dataStream = new MemoryStream(messageLength))
+                {
+                    CopyStream(stream, dataStream, messageLength).GetAwaiter().GetResult();
```

Comment:
Change this line to 
`stream.CopyTo(dataStream, Math.Min(messageLength, 4096));`
and drop `CopyStream` method?